### ISSUE TITLE
Switch Expressions USE: "return match (person))"

### ISSUE THREAD

**Issue by zezba9000** [dotnet/csharplang#1555]
Regarding this topic on odd "switch" statements: https://www.infoq.com/news/2018/05/CSharp-8-Future

Please make it more in line with the C# Style (not some merged in hack from another lang)
I think something like the code below would be a cleaner approach. Use an explicit "match" instead of "switch" as they expect different syntax internally. Making keywords implicit in C# is going to end up hurting the language big time.
```
return match (person)
{
   case Professor: $"Dr. {person.LastName}";
};
```

I don't normally have many issues with C# but this just really bothers me lol

**Comment by HaloFour** [comment]
C# style is that `statement (expression)` is not an expression but a statement and is not expected to return a value.  The syntax you've suggested was in the very long list of potential syntaxes suggested and tried and an inline keyword operator (like `as`) was considered to look and feel better.  That said, this is still under development and they're still attempting different flavors of the syntax.

**Comment by yaakov-h** [comment]
What do you mean by "Making keywords implicit"? What's implicit about the `switch` in `foo switch {`?

**Comment by zezba9000** [comment]
@yaakov-h 

> What do you mean by "Making keywords implicit"? What's implicit about the switch in foo switch {?

The same keyword "switch" now becomes context sensitive. Its use case is implicit on context instead of it explicitly stating the context it expects you to provide.

@HaloFour 
C# style is that statement (expression) is not an expression but a statement and is not expected to return a value. 

If that held true, then why does "return new Person()" expect a return value? The "match" keyword would be considered as one expecting a return value, just as new does. Also "statement (expression)" doesn't hold true in the suggested method of "return person switch" [thats identifier (statement)]. Am I wrong about that?

This is a similar argument for having "block" keyword come before anonymous scopes. In short prefixing the intention of what is to follow seems logical to me.
```
block
{
   // do stuff...
}
```

**Comment by HaloFour** [comment]
@zezba9000 

> The same keyword "switch" now becomes context sensitive.

`switch` is already a reserved keyword.  Why does it matter that you can now use it in two similar contexts?  If anything it's more cohesive than `using` which has three different uses and two very different syntactic forms.

On the other hand, you're suggesting making `match` a contextual keyword, in that the compiler will have to parse ahead (beyond the expression) before the compiler could determine whether or not it is actually a keyword or an identifier of a function or delegate instance.

> If that held true, then why does "return new Person()" expect a return value?

`return` is a control statement.  It *requires* an expression (at least where the function returns a value) but it itself is *not* an expression.  That's why the following is not legal:

```cs
int x = return (1 + 1);
```

> In short prefixing the intention of what is to follow seems logical to me.

All binary operators in C# are infix.  `switch` in this case is an operator where the second operand is a pattern.  This makes it very similar to the existing `is` and `as` operators, which are both infix.

**Comment by zezba9000** [comment]
> more cohesive than using which has three different uses and two very different syntactic forms

C# is my lang of choice but has some major syntax flaws. I never did like that as it confuses so many people. Should have been called "auto" or something for auto calling dispose.

> On the other hand, you're suggesting making match a contextual keyword

No I'm not suggesting it be contextual. I'm suggesting it be reserved, just as "switch", "if", "else", "while", "do", etc are. Just as those control logic flow, the same can be said for "match".

To take this even further, a compare statement in my mind should have a reserved "cp" prefix as well. Example: "int i = cp (0 == 1) ? 33 : 66;". Just as an "if" statement does ect. These explicit prefix keywords tell you and the compiler what to expect down the line with very little parsing. It also tells you and compiler what to look for when reporting issues and trying to fix them.

The more implicit "using" style additions that get added to C# the more messy it gets and the harder it becomes to work with others. **I think a MUCH needed feature in Roslyn are "experimental" features. We need a flag in VS to enable these so they go through longer testing cycles instead of just adding stuff in and hoping its good long term.** Move C# forward but be safer in the direction it takes with longer test cycles. Maybe I'll suggest that on the Roslyn GitHub page.

**Comment by CyrusNajmabadi** [comment]
> No I'm not suggesting it be contextual. I'm suggesting it be reserved

Well, we can't reserve the 'match' identifier as that could break tons of code out htere. In roslyn alone i found many instance of 'match' being used as a name of something.

When this happens, we usually decide on making an identifier a 'contextual keyword'.  i.e. it is fine as an identifier elsewhere, but if we see it with a certain syntactic pattern, then we'll treat it as a keyword.  For example, you can write:

```c#
int from;
int yield;
```

And those are legal.  However, if you write:

```c#
var q = from x ...
```

Then we know *here* that *this* 'from' is a keyword because it is only in a syntactic form that would be legal if it were treated that way.

**Comment by CyrusNajmabadi** [comment]
> Just as those control logic flow, the same can be said for "match".

Match isn't really control-flow in the traditional sense.  It's much more akin to the ternary operator.  An expression which conditionally evaluates a branch based on a value.  

> These explicit prefix keywords tell you and the compiler what to expect down the line with very little parsing

The problem here is that ship sailed long ago.  You're not only complaining about decisions with C# since C# 1.0, you're complaining about decisions from the entire lineage of C-style languages.  We're well past litigating those decisions.  If you want to use languages that have fundamentally different views on keywords and parsing decisions, you should def use those.  But those decisions don't fit into C#'s views on the world.

>  I think a MUCH needed feature in Roslyn are "experimental" features. We need a flag in VS to enable these

C# already has this.  We call them "feature flags" and we hide features behind them until they are ready :)

>  so they go through longer testing cycles instead of just adding stuff in and hoping its good long term. 

Features get very long testing cycles.  And lots of tweaks and changes are pulled into the design of things.  You may not like the outcomes here, but overall your views don't seem to necessarily match those of the larger community and ecosystem out there.  

> I never did like that as it confuses so many people.  ... the more messy it gets 

Then you will always be disappointed :)  .  There isn't a programming language out there that doesn't confuse a heck of a lot of users.  On the scale of languages out there, C# is positively tame in terms of confusing people.  So many other languages have had far less thought and effort put forth into evolving it in a cohesive and sensible manner :)   

As it happens, the way we've chosen this may not be the way you like.  But it's a way that is actually a pretty consistently reliable way to accomplish the goals of moving the language forward, while also trying to keep things feeling like it is the same language.

**Comment by Neme12** [comment]
> The same keyword "switch" now becomes context sensitive. Its use case is implicit on context instead of it explicitly stating the context it expects you to provide.

It already *is* context sensitive.
```c#
switch (value) { } // here it's part of a switch statement
```
```c#
value switch { } // here it doesn't mean anything
```

**Comment by yaakov-h** [comment]
@Neme12 In what sort of code is `value switch { }` currently valid C#?

**Comment by svick** [comment]
@CyrusNajmabadi 

> C# already has this. We call them "feature flags" and we hide features behind them until they are ready :)

One thing is having feature flags, another is actually using them. According to [Language Feature Status](https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md), there are currently 10 features in active development. Which of those can I try by enabling a feature flag?

And even when you want people to try out features that are in development, you don't use feature flags, you use separately installed previews (see https://github.com/dotnet/roslyn/issues/26741 for some issues I have with that model).

I understand that from an engineering perspective, using feature flags is not a trivial thing to do, especially with Roslyn's approach of every feature being developed in its own branch. And not using feature flags might even be the right choice. But it does mean that most people can only get their hands on features once they are almost done, which in effect means that Roslyn does not have experimental features.

**Comment by zezba9000** [comment]
@CyrusNajmabadi 

> Then you will always be disappointed :) . There isn't a programming language out there that doesn't confuse a heck of a lot of users.

Not to be argumentative but this is a horrible argument for bad designs (it suggest we shouldn't care about learning from design mistakes because someone will always "feel" something we don't). Also I'm not always disappointed but am about some aspects of C#. There are reasons its my lang of choice even if I have some complaints.


> You may not like the outcomes here, but overall your views don't seem to necessarily match those of the larger community and ecosystem out there.

The C# community or the community from other langs? I prefer objective arguments rather than subjective ones. Mobs don't always make the best rules.

> C# already has this. We call them "feature flags" and we hide features behind them until they are ready :)

Still would be nice to have a simple VS 2017 flag tab/page (like Chrome does). I think the majority just use stuff after its already released and potential use cases might be missed (just a guess).

> You're not only complaining about decisions with C# since C# 1.0, you're complaining about decisions from the entire lineage of C-style languages.

Yes and no maybe. A prefix thats reserved is required in some aspects of C#. Such as "var" explicitly expects to infer down the line its type. That keyword as it is now doesn't change based on context. Its meaning is constant and unambiguous.

In either case I'll try to just leave this be. I would like to see a more robust "feature flag" tab in VS though. I'll look into it though.

**Comment by HaloFour** [comment]
@zezba9000 

> Not to be argumentative but this is a horrible argument / excuse for bad designs

You not liking particular syntax doesn't make it a bad design.

> I prefer objective arguments rather than subjective ones.

Stylistic debates are entirely subjective.

> That keyword as it is now doesn't change based on context. Its meaning is constant and unambiguous.

`var` does change based on context.  It could be a type:

```cs
public struct var { }

static class Program {
    static void Main() {
        var x = 123; // compiler error
    }
}
```

As a keyword `var` does mean multiple things.  For example it is also a variable pattern.

**Comment by zezba9000** [comment]
@HaloFour 

> You not liking particular syntax doesn't make it a bad design.

That was kind of the point...

> Stylistic debates are entirely subjective.

Not if the argument is about keeping a style consistent (which is the point). Thats objective.

> var does change based on context. It could be a type:

I think you missed QUTOE: "explicitly expects to infer down the line its type." The actually meaning of "var" itself didn't change because of the context. "var" is just syntax sugar for an inferred type.

**Comment by zezba9000** [comment]
> You don't think that the current design in the prototype fits with the rest of the language, but it's built entirely on syntactic aspects of the language that currently exist.

That by definition would be an objective argument if true. You're contradicting yourself.

> I fail to see how that statement is relevant. If your claim that var doesn't change contextually

"var" as a syntactical concept doesn't change depending on context (it always means infer its suffix's type). Var isn't an object itself and thus its meaning doesn't change based on the object it infered. Just because the type it inferred can be anything doesn't mean the actual meaning syntactically behind var changed. Not sure I can be more clear.

**Comment by Neme12** [comment]
> "var" as a syntactical concept doesn't change depending on context

What does that mean? `switch` as a syntactical concept doesn't change either - it's always the `switch` keyword.

**Comment by Neme12** [comment]
> Just because the type it inferred can be anything doesn't mean the actual meaning syntactically behind var changed

The syntax didn't change, but the semantics did, which is a much more important distinction.

**Comment by Neme12** [comment]
> it always means infer its suffix's type

No it doesn't. (by suffix I assume you mean initializer?)
If there's an existing typed named `var` in scope, it's just that. The type `var` and no inference.

**Comment by zezba9000** [comment]
@Neme12 
Seems like you're kinda cherry picking things I've said slightly out of context. I'm closing this.

**Comment by Neme12** [comment]
I'm sorry but I don't understand any points you were trying to make.

**Comment by CyrusNajmabadi** [comment]
> Not if the argument is about keeping a style consistent (which is the point). Thats objective.
> I prefer objective arguments rather than subjective ones.

Ok.  here's an objective argument about keeping style consistent.  C# has precedence for using infix-keywords for binary expression operations on values.  i.e.  ```x is Y```  or ```x as Y``` or ```x is Some { Recursive: var Pattern }```.   Because of that, ```switch``` objectively works as an infix operator that fits with how we've done things in many other cases.

**Comment by CyrusNajmabadi** [comment]
> "var" as a syntactical concept doesn't change depending on context 

THat's actually incorrect.   'var' is specifically *not* syntactic.  It's entirely semantic.  You can see this by looking at the C# parser (the syntactic side of hte compiler) which has absolutely *no* knowledge of ```var```.  (It also has no knowledge of ```dynamic``` or ```unmanaged```).  All of these are processed in an entirely semantic fashion only once the semantic side of hte compiler has determined that these 'words' dont' mean anything else.

In the IDE they show up as 'blue', but only because the IDE sees that these are words that *semantically* (not syntactically) are special.  We decided on 'blue' because that's our general approach for coloring things that are special, but which are themselves not user-defined (like classes/structs/etc.)

I hope that helps clear things up :)

**Comment by Neme12** [comment]
@CyrusNajmabadi That is true but the switch expression is not a binary expression.

**Comment by CyrusNajmabadi** [comment]
I didn't say it was a binary expression :)   Just that's it's an operation on a value.  'is' is not really binary either.  It either has a 'Type' on the right (not an expression), or it has a pattern on the right (also not an expression) :)

**Comment by CyrusNajmabadi** [comment]
regardless, 'switch' or 'match' feel completely fine on the right.  *And* they wouldo also feel fine on the left (to me) where they would be similar to things like "nameof"/"typeof" etc.  We have lots of options here, and it's ultimately going to just be one the LDM feels best about, and will invariably have people who dislike it.

**Comment by Neme12** [comment]
I thought types were expressions by the language spec

**Comment by CyrusNajmabadi** [comment]
> I thought types were expressions by the language spec

I don't believe that's the case :)    We do allow that in the actual impl for ease of error correction and error modeling.  But that's not how the language actually sees things.

**Comment by Neme12** [comment]
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#expression-classifications
> A type. An expression with this classification can only appear as the left hand side of a member_access (Member access), or as an operand for the as operator (The as operator), the is operator (The is operator), or the typeof operator (The typeof operator). In any other context, an expression classified as a type causes a compile-time error.

**Comment by Neme12** [comment]
Anyway, let's not argue semantics. I just thought that was interesting but wouldn't think of types as expressions myself. And you're right about patterns.

**Comment by CyrusNajmabadi** [comment]
> Anyway, let's not argue semantics.

![image](https://user-images.githubusercontent.com/4564579/40446956-f87b2df4-5e9e-11e8-8c3a-cbe052d43177.png)

:D

Btw, i have some questions for you if you're available over on gitter.  Stop by if you have a minute!

**Comment by zezba9000** [comment]
@CyrusNajmabadi 

> I hope that helps clear things up :)

Somewhat but just feels wrong I can do "int var = 0;" but NOT "int if = 0;" When people write code most I would assume consider this an inconstancy in the language. Language (even spoken) start to become hard to communicate complex idea's when they're full of "gotcha's" for lack of a better word. When I look at syntax thats kinda where my mind goes when considering something one method over another... if that makes sense.

**Comment by jnm2** [comment]
@zezba9000 I don't think that gotcha is important. I'd reject `int var = 0;` in a pull request.

**Comment by CyrusNajmabadi** [comment]
> Somewhat but just feels wrong I can do "int var = 0;" but NOT "int if = 0;"

Such is hte nature of back compat.  We can't add keywords to the language because it could break existing code.  Clearly that wasn't a concern for 1.0 :)     So things like 'if' are just full 'reserved' words.  Things like 'var' are just purely a semantic addition.

Now, with 'switch' we're lucky.  It's already a keyword.  So no one could ever use it for any other purpose.  That means we can just create whatever sort of syntactic constructs we want with it and not have to worry about colliding with anything else.  W00t!

> When people write code most I would assume consider this an inconstancy in the language. 

Most people don't even think about it :)   Which is a good thing.  :)

Note: there is no real way to avoid this unless you want to potentially break existing code.  It's a core goal of C# to not do that.  So this is what you get.  :)
